src/node_watchdog.cc

```diff
@@ -100,16 +124,55 @@ SigintWatchdog::SigintWatchdog(
   SigintWatchdogHelper::GetInstance()->Start();
 }
 
+SigintWatchdog::SigintWatchdog(
+  v8::Isolate* isolate, Local<v8::Function> interrupt)
+    : isolate_(isolate) {
+  ref.Reset(isolate, interrupt);
+  auto env = node::Environment::GetCurrent(isolate);
+  loop_activate = new uv_async_t;
+  loop_activate->data = this;
+  // Just activate the loop, doesn't do callback.
+  uv_async_init(env->event_loop(), loop_activate, [](uv_async_t* it){
+    SigintWatchdog* self = reinterpret_cast<SigintWatchdog*>(it->data);
+    self->HandleInterrupt();
+  });
+  uv_unref(reinterpret_cast<uv_handle_t*>(loop_activate));
+  // Register this watchdog with the global SIGINT/Ctrl+C listener.
+  SigintWatchdogHelper::GetInstance()->Register(this);
+  // Start the helper thread, if that has not already happened.
+  SigintWatchdogHelper::GetInstance()->Start();
+}
 
 SigintWatchdog::~SigintWatchdog() {
+  if (loop_activate != nullptr) {
+    uv_close(reinterpret_cast<uv_handle_t*>(loop_activate),
```

Comment:
You’d want to use `env->CloseHandle(...)` rather than `uv_close()` here, so that the Environment can keep track of handles that are still closing, but that problem goes away if you switch to using `HandleWrap`.